src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java

```diff
@@ -99,19 +105,137 @@ else if(agent.getDockerfile() != null){
 
                 agentLines.add(getCommonOptionsOfAgent(agent));
                 agentLines.add("}");
+                agentLines.add("}");
             }
-            else {
+            else if(agent.getLabel() != null && agent.getCustomWorkspace() != null){
+                agentLines.add("agent {");
                 agentLines.add("node{");
                 agentLines.add(getCommonOptionsOfAgent(agent));
                 agentLines.add("}");
+                agentLines.add("}");
+            }
+            else {
+                agentLines.add("agent any");
             }
 
-            agentLines.add("}");
+        }
+
+        if(agent != null) {
+            agentLines.addAll(getTools(agent.getTools()));
         }
 
         return agentLines;
     }
 
+    public List<String> getTools(HashMap<String, String> tools){
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(tools == null) {
+            return snippetLines;
+        }
+
+        snippetLines.add("tools {");
+
+        for(Map.Entry<String, String> entry: tools.entrySet()){
+            snippetLines.add(entry.getKey() + " '" + entry.getValue() + "'");
+        }
+
+        snippetLines.add("}");
+
+        return snippetLines;
+    }
+
+    public List<String> getEnvironment(Environment environment){
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(environment == null || (environment.getVariables() == null && environment.getCredentials() == null)) {
+            return snippetLines;
+        }
+
+        snippetLines.add("environment {");
+
+        for(Map.Entry<String, String> entry: environment.getVariables().entrySet()){
+            snippetLines.add(entry.getKey() + " = '" + entry.getValue() + "'");
+        }
+
+        for(Credential credential: environment.getCredentials()){
+            snippetLines.add(credential.getVariable() + " = credentials('" + credential.getCredentialId() + "')");
+        }
+
+        snippetLines.add("}");
+
+        return snippetLines;
+    }
+
+    public List<String> getPostSection(Post postSection) throws NoSuchMethodException, ConfiguratorException,
+            InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchFieldException {
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(postSection == null) {
+            return snippetLines;
+        }
+
+        snippetLines.add("post {");
+
+        if(postSection.getAborted() != null){
+            snippetLines.add("aborted {");
+            snippetLines.addAll(getSteps(postSection.getAborted()));
+            snippetLines.add("}");
+        }
+
+        if(postSection.getAlways() != null){
```

Comment:
Apart from code style, some bits of code deduplication could be applied here, e.g. by introducing a method: like 

```java
private void addPostIfNonNull(@Nonnull StringBuilder snippetLines , @Nonnull String condition, @CheckfForNull Whatever content)
        if(content != null) {
            snippetLines.add(condition + " {");
            snippetLines.addAll(getSteps(content));
            snippetLines.add("}");
        }
  }
```